### ISSUE TITLE
TestWatchErrConnClosed fails

### ISSUE THREAD

**Issue by gyuho** [coreos/etcd#6630]
https://semaphoreci.com/coreos/etcd/branches/master/builds/1076

[log.txt](https://github.com/coreos/etcd/files/523189/log.txt)

```
=== RUN   TestWatchErrConnClosed
2016-10-11 23:20:02.040819 I | integration: launching node2522201834049211672 (unix://localhost:node2522201834049211672.sock.bridge)
2016-10-11 23:20:02.041431 I | etcdserver: name = node2522201834049211672
2016-10-11 23:20:02.041471 I | etcdserver: data dir = /tmp/etcd506766226
2016-10-11 23:20:02.041512 I | etcdserver: member dir = /tmp/etcd506766226/member
2016-10-11 23:20:02.041550 I | etcdserver: heartbeat = 10ms
2016-10-11 23:20:02.041582 I | etcdserver: election = 100ms
2016-10-11 23:20:02.041611 I | etcdserver: snapshot count = 0
2016-10-11 23:20:02.041649 I | etcdserver: advertise client URLs = unix://127.0.0.1:21524.16121.sock
2016-10-11 23:20:02.041688 I | etcdserver: initial advertise peer URLs = unix://127.0.0.1:21523.16121.sock
2016-10-11 23:20:02.041736 I | etcdserver: initial cluster = node2522201834049211672=unix://127.0.0.1:21523.16121.sock
2016-10-11 23:20:02.060916 I | etcdserver: starting member 2e4b02ca96375a21 in cluster 97a4560d2d323a5f
2016-10-11 23:20:02.060995 I | raft: 2e4b02ca96375a21 became follower at term 0
2016-10-11 23:20:02.061059 I | raft: newRaft 2e4b02ca96375a21 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016-10-11 23:20:02.061104 I | raft: 2e4b02ca96375a21 became follower at term 1
2016-10-11 23:20:02.062580 I | etcdserver: set snapshot count to default 10000
2016-10-11 23:20:02.062623 I | etcdserver: starting server... [version: 3.1.0-alpha.1+git, cluster version: to_be_decided]
2016-10-11 23:20:02.081248 I | etcdserver/membership: added member 2e4b02ca96375a21 [unix://127.0.0.1:21523.16121.sock] to cluster 97a4560d2d323a5f
2016-10-11 23:20:02.089753 I | integration: launched node2522201834049211672 (unix://localhost:node2522201834049211672.sock.bridge)
2016-10-11 23:20:02.089834 I | raft: 2e4b02ca96375a21 is starting a new election at term 1
2016-10-11 23:20:02.089904 I | raft: 2e4b02ca96375a21 became candidate at term 2
2016-10-11 23:20:02.089945 I | raft: 2e4b02ca96375a21 received vote from 2e4b02ca96375a21 at term 2
2016-10-11 23:20:02.090001 I | raft: 2e4b02ca96375a21 became leader at term 2
2016-10-11 23:20:02.090046 I | raft: raft.node: 2e4b02ca96375a21 elected leader 2e4b02ca96375a21 at term 2
2016-10-11 23:20:02.096163 W | etcdserver: apply entries took too long [13.066724ms for 1 entries]
2016-10-11 23:20:02.097919 W | etcdserver: avoid queries with large range/delete range!
2016-10-11 23:20:02.098634 I | etcdserver: published {Name:node2522201834049211672 ClientURLs:[unix://127.0.0.1:21524.16121.sock]} to cluster 97a4560d2d323a5f
2016-10-11 23:20:02.098998 I | etcdserver: setting up the initial cluster version to 3.1
2016-10-11 23:20:02.099473 N | etcdserver/membership: set the initial cluster version to 3.1
2016-10-11 23:20:02.115156 I | integration: terminating node2522201834049211672 (unix://localhost:node2522201834049211672.sock.bridge)
2016-10-11 23:20:02.144478 I | integration: terminated node2522201834049211672 (unix://localhost:node2522201834049211672.sock.bridge)
--- FAIL: TestWatchErrConnClosed (0.11s)
    watch_test.go:589: expected grpc: the client connection is closing, got context canceled
```

**Comment by gyuho** [comment]
@heyitsanthony @xiang90 

In what cases do we expect `ErrClientConnClosing` from closing watch client?

I've been running this `TestWatchErrConnClosed` https://github.com/coreos/etcd/blob/master/clientv3/integration/watch_test.go#L610 over 5,000 rounds but cannot reproduce.

```go
func TestWatchError(t *testing.T) {
	defer testutil.AfterTest(t)

	clus := integration.NewClusterV3(t, &integration.ClusterConfig{Size: 1})
	defer clus.Terminate(t)

	cli := clus.Client(0)
	wc := clientv3.NewWatcher(cli)

	cerr := cli.Close()
	fmt.Println("cli.Close:", cerr)

	wc.Watch(context.TODO(), "foo")
	err := wc.Close()
	fmt.Println("wc.Close:", err)

	clus.TakeClient(0)
}
```

This both returns nil. `TestWatchErrConnClosed` always returns `nil` error.

Am I missing anything?

Thanks!

**Comment by heyitsanthony** [comment]
It should happen if the grpc connection is closed before the watch client. `clientv3.Close` won't trigger it any more because it's closing the watch client now instead of only closing the grpc connection.

This should do it:
```go
conn := cli.ActiveConnection()
conn.Close()
wr <- wc
// wr.Err() == closing
```

**Comment by gyuho** [comment]
@heyitsanthony How about changing the test as

```diff
diff --git a/clientv3/integration/watch_test.go b/clientv3/integration/watch_test.go
index 1b544ca..4207253 100644
--- a/clientv3/integration/watch_test.go
+++ b/clientv3/integration/watch_test.go
@@ -616,18 +616,21 @@ func TestWatchErrConnClosed(t *testing.T) {
        cli := clus.Client(0)
        wc := clientv3.NewWatcher(cli)
 
-       donec := make(chan struct{})
+       donec, readyc := make(chan struct{}), make(chan struct{})
        go func() {
                defer close(donec)
-               wc.Watch(context.TODO(), "foo")
-               if err := wc.Close(); err != nil && err != grpc.ErrClientConnClosing {
-                       t.Fatalf("expected %v, got %v", grpc.ErrClientConnClosing, err)
+               <-readyc
+               ch := wc.Watch(context.TODO(), "foo")
+               wc.Close()
+               if wr := <-ch; grpc.ErrorDesc(wr.Err()) != grpc.ErrClientConnClosing.Error() {
+                       t.Fatalf("expected %v, got %v", grpc.ErrClientConnClosing, grpc.ErrorDesc(wr.Err()))
                }
        }()
 
-       if err := cli.Close(); err != nil {
+       if err := cli.ActiveConnection().Close(); err != nil {
                t.Fatal(err)
        }
+       close(readyc)
        clus.TakeClient(0)
```

Currently it won't return `ErrConnClosing` as you said.

I am still not sure how we got the context error. Will try more to reproduce.

**Comment by heyitsanthony** [comment]
@gyuho There's no running watchers in that example since the `readyc` stuff clobbers the concurrency between closing the client and creating a watch + closing the watch client. The watcher client will never see the closing error because it never tries to establish grpc stream.

**Comment by gyuho** [comment]
Ok got it. Thanks.

**Comment by gyuho** [comment]
Closing via https://github.com/coreos/etcd/pull/6768.

Will reopen if it happens again.

**Comment by heyitsanthony** [comment]
via https://jenkins-etcd-public.prod.coreos.systems/job/etcd-proxy/761/console
```
=== RUN   TestWatchErrConnClosed
2017-02-24 23:58:32.526682 I | integration: launching 4098451428066286304 (unix://localhost:40984514280662863040)
2017-02-24 23:58:32.532204 I | etcdserver: name = 4098451428066286304
2017-02-24 23:58:32.532292 I | etcdserver: data dir = /tmp/etcd959722066
2017-02-24 23:58:32.532465 I | etcdserver: member dir = /tmp/etcd959722066/member
2017-02-24 23:58:32.532532 I | etcdserver: heartbeat = 10ms
2017-02-24 23:58:32.532641 I | etcdserver: election = 100ms
2017-02-24 23:58:32.532768 I | etcdserver: snapshot count = 0
2017-02-24 23:58:32.532947 I | etcdserver: advertise client URLs = unix://127.0.0.1:2190410462
2017-02-24 23:58:32.533057 I | etcdserver: initial advertise peer URLs = unix://127.0.0.1:2190310462
2017-02-24 23:58:32.533201 I | etcdserver: initial cluster = 4098451428066286304=unix://127.0.0.1:2190310462
2017-02-24 23:58:32.539270 I | etcdserver: starting member 853d55cc9f58ff55 in cluster 1ac23b84f2ded41
2017-02-24 23:58:32.539504 I | raft: 853d55cc9f58ff55 became follower at term 0
2017-02-24 23:58:32.539657 I | raft: newRaft 853d55cc9f58ff55 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2017-02-24 23:58:32.539734 I | raft: 853d55cc9f58ff55 became follower at term 1
2017-02-24 23:58:32.557636 I | etcdserver: set snapshot count to default 100000
2017-02-24 23:58:32.557762 I | etcdserver: starting server... [version: 3.2.0+git, cluster version: to_be_decided]
2017-02-24 23:58:32.560362 I | raft: 853d55cc9f58ff55 is starting a new election at term 1
2017-02-24 23:58:32.560620 I | raft: 853d55cc9f58ff55 became candidate at term 2
2017-02-24 23:58:32.560788 I | raft: 853d55cc9f58ff55 received MsgVoteResp from 853d55cc9f58ff55 at term 2
2017-02-24 23:58:32.560956 I | raft: 853d55cc9f58ff55 became leader at term 2
2017-02-24 23:58:32.561386 I | etcdserver/membership: added member 853d55cc9f58ff55 [unix://127.0.0.1:2190310462] to cluster 1ac23b84f2ded41
2017-02-24 23:58:32.561759 I | raft: raft.node: 853d55cc9f58ff55 elected leader 853d55cc9f58ff55 at term 2
2017-02-24 23:58:32.565183 I | etcdserver: setting up the initial cluster version to 3.2
2017-02-24 23:58:32.566359 I | etcdserver: published {Name:4098451428066286304 ClientURLs:[unix://127.0.0.1:2190410462]} to cluster 1ac23b84f2ded41
2017-02-24 23:58:32.567303 N | etcdserver/membership: set the initial cluster version to 3.2
2017-02-24 23:58:32.570859 I | integration: launched 4098451428066286304 (unix://localhost:40984514280662863040)
2017-02-24 23:58:32.588035 I | integration: terminating 4098451428066286304 (unix://localhost:40984514280662863040)
2017-02-24 23:58:32.597469 I | integration: terminated 4098451428066286304 (unix://localhost:40984514280662863040)
--- FAIL: TestWatchErrConnClosed (0.07s)
	watch_test.go:616: expected grpc: the client connection is closing, got context canceled
```

May be unique to the grpcproxy or the timing is nudged just enough to trigger it.
